File: checker/tests/test_hooks_tasks.py
"""Public checker seam tests for the deterministic Hooks/Tasks lesson."""

from __future__ import annotations

import json
import os
import subprocess
import sys
import tempfile
import unittest
from pathlib import Path


ROOT = Path(__file__).resolve().parents[2]
CHECKER = ROOT / "checker"
COURSE_VERSION = json.loads(
    (ROOT / "course-version.json").read_text(encoding="utf-8")
)["course_version"]


def passed_fixture() -> dict[str, object]:
    return {
        "contract": "agent-engineering-course/evidence",
        "contract_version": "1",
        "course_version": COURSE_VERSION,
        "lesson_id": "t21-hooks-tasks",
        "result": "passed",
        "anonymous": True,
        "checked_on": "2026-08-13",
        "summary": "所有必需证据均已通过。",
        "evidence": [
            {"id": "trigger-observed", "result": "passed"},
            {"id": "deduplication-observed", "result": "passed"},
            {"id": "permission-boundary", "result": "passed"},
            {"id": "stop-condition", "result": "passed"},
            {"id": "failure-recovered", "result": "passed"},
            {"id": "side-effect-not-triggered", "result": "passed"},
            {"id": "explicit-task-recorded", "result": "passed"},
            {"id": "offline-deterministic", "result": "passed"},
        ],
        "experiment": {
            "version": "1",
            "runs": [
                {
                    "id": "run-1",
                    "mode": "hook",
                    "trigger": True,
                    "deduplicated": True,
                    "permission": "blocked",
                    "stopped": True,
                    "failed": True,
                    "recovered": True,
                    "taskCreated": True,
                    "sideEffect": False,
                    "scheduleArmed": False,
                    "backgroundStarted": False,
                }
            ],
            "observed": [
                "deduplication",
                "explicit-task",
                "failure-recovery",
                "permission",
                "side-effect-guard",
                "stop",
                "trigger",
            ],
        },
    }


class HooksTasksCheckerTests(unittest.TestCase):
    def run_checker(self, *args: str) -> subprocess.CompletedProcess[str]:
        env = os.environ.copy()
        env["PYTHONPATH"] = str(CHECKER)
        return subprocess.run(
            [sys.executable, "-m", "course_check", *args],
            cwd=CHECKER,
            env=env,
            capture_output=True,
            text=True,
            encoding="utf-8",
            check=False,
        )

    def test_structure_check_is_partial_without_learner_evidence(self) -> None:
        result = self.run_checker("check", "t21-hooks-tasks", "--root", str(ROOT), "--json")
        self.assertEqual(result.returncode, 0, result.stderr)
        document = json.loads(result.stdout)
        self.assertEqual(document["result"], "partial")
        self.assertEqual(document["lesson_id"], "t21-hooks-tasks")

    def test_checker_derives_passed_result_from_anonymous_runs(self) -> None:
        with tempfile.TemporaryDirectory() as temporary:
            fixture = Path(temporary) / "t21-hooks-tasks-evidence.json"
            fixture.write_text(json.dumps(passed_fixture(), ensure_ascii=False), encoding="utf-8")
            result = self.run_checker(
                "check",
                "t21-hooks-tasks",
                "--root",
                str(ROOT),
                "--evidence-file",
                str(fixture),
                "--json",
            )
        self.assertEqual(result.returncode, 0, result.stderr)
        document = json.loads(result.stdout)
        self.assertEqual(document["result"], "passed")
        self.assertNotIn(str(fixture), result.stdout)
        self.assertNotIn("command", result.stdout)

    def test_checker_rejects_forged_checks_that_do_not_match_runs(self) -> None:
        fixture = passed_fixture()
        experiment = fixture["experiment"]
        assert isinstance(experiment, dict)
        runs = experiment["runs"]
        assert isinstance(runs, list)
        runs[0]["recovered"] = False
        with tempfile.TemporaryDirectory() as temporary:
            path = Path(temporary) / "forged.json"
            path.write_text(json.dumps(fixture, ensure_ascii=False), encoding="utf-8")
            result = self.run_checker(
                "check",
                "t21-hooks-tasks",
                "--root",
                str(ROOT),
                "--evidence-file",
                str(path),
                "--json",
            )
        self.assertNotEqual(result.returncode, 0)
        self.assertIn("T21", result.stderr)

    def test_checker_rejects_side_effect_without_allowed_permission(self) -> None:
        fixture = passed_fixture()
        experiment = fixture["experiment"]
        assert isinstance(experiment, dict)
        runs = experiment["runs"]
        assert isinstance(runs, list)
        runs[0]["sideEffect"] = True
        with tempfile.TemporaryDirectory() as temporary:
            path = Path(temporary) / "unsafe.json"
            path.write_text(json.dumps(fixture, ensure_ascii=False), encoding="utf-8")
            result = self.run_checker(
                "check",
                "t21-hooks-tasks",
                "--root",
                str(ROOT),
                "--evidence-file",
                str(path),
                "--json",
            )
        self.assertNotEqual(result.returncode, 0)
        self.assertIn("side effect", result.stderr)


if __name__ == "__main__":
    unittest.main()

File: site/scripts/test-hooks-tasks.mjs
import assert from "node:assert/strict";
import test from "node:test";

import {
  DEFAULT_HOOKS_TASKS_INPUT,
  HOOKS_TASKS_PRESETS,
  buildHooksTasksEvidence,
  createHooksTasksSession,
  recordHooksTasksObservation,
  simulateHooksTasks,
  updateHooksTasksSession,
} from "../src/lib/hooks-tasks.mjs";

test("safe default deterministically covers trigger, deduplication, recovery and stop", () => {
  const first = simulateHooksTasks(DEFAULT_HOOKS_TASKS_INPUT);
  const second = simulateHooksTasks(DEFAULT_HOOKS_TASKS_INPUT);
  assert.deepEqual(first, second);
  assert.equal(first.run.triggered, true);
  assert.equal(first.run.deduplicated, 1);
  assert.equal(first.run.permission, "blocked");
  assert.equal(first.run.sideEffect, false);
  assert.equal(first.run.failureInjected, true);
  assert.equal(first.run.recovered, true);
  assert.equal(first.run.stopped, true);
  assert.equal(first.run.explicitTask, true);
});

test("schedule and background are not silently treated as automatic hooks", () => {
  for (const name of ["schedule", "background"]) {
    const result = simulateHooksTasks(HOOKS_TASKS_PRESETS[name]).run;
    assert.equal(result.triggered, false);
    assert.equal(result.sideEffect, false);
    assert.equal(name === "schedule", result.scheduleArmed);
    assert.equal(name === "background", result.backgroundStarted);
  }
});

test("permission and failure presets expose risk and stopping branches", () => {
  const allowed = simulateHooksTasks(HOOKS_TASKS_PRESETS.permission).run;
  assert.equal(allowed.permission, "allowed");
  assert.equal(allowed.sideEffect, true);

  const persistent = simulateHooksTasks(HOOKS_TASKS_PRESETS.failure).run;
  assert.equal(persistent.failureUnresolved, true);
  assert.equal(persistent.recovered, false);
  assert.equal(persistent.stopped, true);
  assert.equal(persistent.stopReason, "step-budget");
});

test("evidence contains stable states only and reaches passed with the safe run", () => {
  let session = createHooksTasksSession(DEFAULT_HOOKS_TASKS_INPUT);
  session = recordHooksTasksObservation(session);
  const evidence = buildHooksTasksEvidence(session, {
    courseVersion: "3.0.0",
    checkedOn: "2026-08-13",
  });
  assert.equal(evidence.lesson_id, "t21-hooks-tasks");
  assert.equal(evidence.result, "passed");
  assert.equal(evidence.anonymous, true);
  assert.deepEqual(
    evidence.evidence.map((check) => check.id),
    [
      "trigger-observed",
      "deduplication-observed",
      "permission-boundary",
      "stop-condition",
      "failure-recovered",
      "side-effect-not-triggered",
      "explicit-task-recorded",
      "offline-deterministic",
    ],
  );
  assert.equal("event" in evidence.experiment.runs[0], false);
  assert.equal("command" in evidence.experiment, false);
});

test("recorded observations survive input changes", () => {
  let session = createHooksTasksSession(DEFAULT_HOOKS_TASKS_INPUT);
  session = recordHooksTasksObservation(session);
  session = updateHooksTasksSession(session, HOOKS_TASKS_PRESETS.explicitTask);
  assert.equal(session.runs.length, 1);
  session = recordHooksTasksObservation(session);
  assert.equal(session.runs.length, 2);
  assert.equal(session.runs[1].taskCreated, true);
});

test("invalid input and unsafe evidence boundaries are rejected", () => {
  assert.throws(
    () => simulateHooksTasks({ ...DEFAULT_HOOKS_TASKS_INPUT, maxSteps: 0 }),
    (error) => error.code === "invalid-maxSteps",
  );
  assert.throws(
    () => simulateHooksTasks({ ...DEFAULT_HOOKS_TASKS_INPUT, mode: "unknown" }),
    (error) => error.code === "invalid-mode",
  );
});
